File: src/main/java/at/yawk/fiction/Image.java
package at.yawk.fiction;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import java.net.URI;
import javax.annotation.Nullable;
import lombok.Data;

/**
 * @author yawkat
 */
@Data
@JsonTypeInfo(use = JsonTypeInfo.Id.MINIMAL_CLASS)
@JsonIgnoreProperties("referer")
public class Image {
    URI imageUrl;
    URI thumbnailUrl;

    public Image() {}

    public Image(URI imageUrl) {
        this.imageUrl = imageUrl;
    }

    public Image(URI imageUrl, URI thumbnailUrl) {
        this.imageUrl = imageUrl;
        this.thumbnailUrl = thumbnailUrl;
    }
}

File: src/main/kotlin/at/yawk/fiction/Image.kt
package at.yawk.fiction

import java.net.URI

/**
 * @author yawkat
 */
data class Image(
        val imageUrl: URI?,
        val thumbnailUrl: URI?,
        /**
         * The HTTP referer required to fetch these images. Note that the header is called `Referer`, this is a
         * misspelling in the HTTP standard.
         */
        @SuppressWarnings("SpellCheckingInspection")
        val referer: URI? = null
) : Mergeable<Image> {
    override fun merge(other: Image): Image {
        return Image(
                imageUrl = imageUrl ?: other.imageUrl,
                thumbnailUrl = thumbnailUrl ?: other.thumbnailUrl,
                referer = referer ?: other.referer
        )
    }
}